File: JAVA/Add 1 to a Linked List Number.java
class Solution {
    public Node addOne(Node head) {
        Node temp = reverse(head);
        Node ans = null; 
        Node node = null; 
        int carry = 1;
        while (temp != null) {
            int sum = temp.data + (carry ==1 ? 1 : 0); 
            carry = sum / 10; 
            sum = sum % 10; 
            Node newNode = new Node(sum);
            if (ans == null) {
                ans = newNode;
                node = newNode;
            } else {
                node.next = newNode;
                node = node.next;
            }
            temp = temp.next;
        }
        if (carry==1) {
            node.next = new Node(1);
        }
        Node fans = reverse(ans);
        return fans;
    }
    public Node reverse(Node head) {
        Node prev = null;
        Node curr = head;
        while (curr != null) {
            Node sec = curr.next;
            curr.next = prev;
            prev = curr;
            curr = sec;
        }
        return prev;
    }
}

File: C++/Add 1 to a Linked List Number.cpp
class Solution {
  public:
    Node* addOne(Node* head) {
        Node* temp = head;
        Node* change = NULL;
        while(temp){
            Node* prev = temp;
            if(temp->next == NULL){
                if(temp->data < 9){
                    temp->data += 1;
                    return head;
                }
                break;
            }
            temp = temp->next;
            if(prev->data!=9 && temp && temp->data==9){
                change = prev;
            }
        }
        if(change==NULL){
            Node* first = new Node(1);
            first->next = head;
            while(head){
                head->data = 0;
                head= head->next;
            }
            return first;
        }
        else{
            change->data += 1;
            change = change->next;
            while(change){
                change->data = 0;
                change= change->next;
            }
            return head;
        }
        return head;
    }
};
